sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java

```diff
@@ -462,6 +488,9 @@ default InputT createInput(Pipeline p, Map<String, PCollection<?>> inputs) {
                         throw new RuntimeException(exn);
                       }
                     }));
+
+    SplittableParDo.convertReadBasedSplittableDoFnsToPrimitiveReadsIfNecessary(pipeline);
```

Comment:
The `convertReadBasedSplittableDoFnsToPrimitiveReadsIfNecessary` already contains this check. Actually, it has the 'canonical' implementation of it - it tests both `use_deprecated_read` and `beam_fn_api_use_deprecated_read` (and can be overridden by `use_sdf_read`). I'd have to reimplement this check or else I would change the semantics.

Reply:
So IMO the default path should be SDF.
All portable runners are expected to support SDF.
Multi-language only works for portable runners.

Flink supporting non-SDF sources is a special case. But if we need to support that it should be behind the experiment "use_deprecated_read". Default path (SDF) should not require an experiment .

Reply:
Sure, agree. The default is SDF. Everything else is triggered by `use_deprecated_read` or similar.

Reply:
But this call "SplittableParDo.convertReadBasedSplittableDoFnsToPrimitiveReadsIfNecessary" being in the default path, it looks like we will convert SDFs to "PrimitiveBoundedRead" if "use_sdf_read" is not set. Hence my request to move this to the else block above.
https://github.com/apache/beam/blob/403ad51d595677d580c7a5cfd9ba9aea6f5793c2/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java#L655

(I understand that we set "use_sdf_read" above for now but removal of that in the future will make default path break for SDF which is required by all runners other than Flink)

Reply:
Oops. My bad. I thought the `if` statement was (!use_sdf_read && (use_deprecated_read || beam_fn_api_use_deprecated_read)) - which actually would make more sense to me, because that is consistent with the fact, that we made SDF the default. This way I think the `beam_fn_api_use_deprecated_read` will not work, but I actually don't know where it is set. Will move the call to the inner block.